### PR TITLE
[Feature] Add fault tolerance with retry and MC2 scale-down on Ascend NPU

### PR DESCRIPTION
**Pull request by fangyuchu**
### What this PR does / why we need it?

In DP+EP (Data Parallelism + Expert Parallelism) MoE deployments, a dead DP rank
makes the EP all2all on surviving ranks block indefinitely, hanging the whole
cluster. Upstream vLLM introduced a fault-tolerance framework
(vllm-project/vllm#44428) that detects the fault, aborts in-flight requests and
lets an external orchestrator trigger coordinated recovery, but Ascend NPU needs
its own backend.

This PR provides the complete Ascend NPU fault-tolerance backend in two layers:

1. **Retry-based recovery for transient faults** (foundation). Reuses the
   upstream `GPUWorkerSentinel` flow (clean worker state + DP-group
   re-initialization on retry) and adds the NPU-specific pieces upstream cannot
   cover: aborts hung NPU communication ops via `torch.npu.npu.set_op_timeout_ms`
   (new `ft_communication_abort_timeout` ascend config) and applies the CPU
   distributed timeout to DP-sync process groups, so peer failures surface as
   exceptions instead of hanging.

2. **Scale-down recovery for MC2** (this PR's focus). When a rank is confirmed
   dead, retry alone cannot recover - survivors keep routing MoE tokens to the
   dead EP rank and the all2all keeps failing. This mirrors the upstream GPU
   `scale_down` flow while adapting the mask, routing and weight layers to
   Ascend's MC2 all2all backend:
   - `npu_communicator`: turns the no-op MC2 all2all manager into the owner of
     the dead-rank mask, encoded as the MC2 `elastic_info` tensor (flag, dense EP
     size, shared/physical expert counts plus orig-to-dense and dense-to-orig rank
     tables), rebuilt in place so captured graphs stay valid.
   - `npu_worker_sentinel`: adds `scale_down` with precondition validation (v2
     model runner, EPLB redundancy, no `fused_mc2`/hierarchy comm, `dispatch_v2`),
     dead EP-rank masking, expert redistribution with routing-table refresh and
     densification, selective on-disk expert weight reload, MC2 physical-expert
     width shrink, and a dummy-batch runnability check.
   - `eplb_redistribute`: NPU expert redistribution + weight reload mirroring
     `process_weights_after_loading` per slot (transpose, FRACTAL_NZ cast, quant
     scales) for unquantized and W8A8 dynamic schemes, plus
     `densify_routing_table_physical_ids`.
   - `token_dispatcher` / `patch_distributed`: pass `elastic_info` to MC2
     dispatch/combine and carry `dead_dp_ranks` so the DP allreduce neutralizes
     dead columns.
   - Clears a poisoned NPU stream on the first device error
     (`fault_barrier_wrapper`) so teardown cannot crash the worker before
     recovery runs.

### Does this PR introduce _any_ user-facing change?

Yes, when `--enable-fault-tolerance` is used on Ascend NPU:

- A new ascend config `ft_communication_abort_timeout` controls how long a hung
  NPU operator (e.g. a communication op) is allowed to run before being aborted
  with a timeout exception. When `> 0`, it derives `HCCL_EVENT_TIMEOUT` /
  `HCCL_EXEC_TIMEOUT` and calls `torch.npu.set_op_timeout_ms` (default `0`,
  disabled).
- Scale-down recovery now supports MC2 deployments and requires EPLB with
  `num_redundant_experts > 0` and the v2 model runner.
- No API/CLI interface changes beyond the upstream FT flags.

Example startup command for a 4-NPU DP=4 Qwen3-30B-A3B-W8A8 deployment (rank 0):

```bash
# Rank 0
ASCEND_RT_VISIBLE_DEVICES=0 \
HCCL_BUFFSIZE=2048 \
HCCL_EXEC_TIMEOUT=1 \
vllm serve /home/xxxxx/Qwen3-30B-A3B-W8A8/ \
  --tensor-parallel-size 1 \
  --data-parallel-size 4 \
  --trust-remote-code \
  --gpu-memory-utilization 0.9 \
  --max-model-len 1024 \
  --enable-expert-parallel \
  --quantization ascend \
  --max-num-batched-tokens 32 \
  --cpu-distributed-timeout-seconds 30 \
  --enable-fault-tolerance \
  --enable-eplb \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE","cudagraph_capture_sizes":[4,8,12,16,20,24,28,32]}' \
  --fault-tolerance-config '{"engine_recovery_timeout_sec":32000}' \
  --eplb-config '{"num_redundant_experts":64, "use_async": false}' \
  --data-parallel-rank 0 \
  --port 8111 \
  --additional-config '{"ft_communication_abort_timeout": 30}' 

# Ranks 1/2/3 use the same command with:
#   ASCEND_RT_VISIBLE_DEVICES={1,2,3}
#   --data-parallel-rank {1,2,3}
#   --port {8112,8113,8114}
```

### How was this patch tested?
Added DP=4 end-to-end coverage in tests/e2e/pull_request/four_card/fault_tolerance/test_fault_tolerance_e2e.py:
-  test_injected_fault_retry_recovers_all_ranks — injected fault on rank 3 drives all 4 engines UNHEALTHY; retry restores all ranks to HEALTHY and factual-prompt serving resumes.
- test_scale_down_removes_dead_rank_and_recovers — SIGKILL of rank 3's worker marks survivors UNHEALTHY and the victim DEAD (DEAD rejects retry); scale_down re-hosts the dead rank's EPLB experts on the survivors, and all survivors return to HEALTHY and answer factual prompts correctly.

vLLM main: https://github.com/vllm-project/vllm/commit/b2f685834a6456197e7033966fdef52a23f1abcd

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/worker/sentinel/npu_worker_sentinel.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        try:
+            return func(self, *args, **kwargs)
+        except SystemExit:
+            raise
+        except Exception as exc:
+            if sentinel is not None:
+                sentinel.worker_faulted = True
+                logger.warning("[FT] Quarantining worker %d after fault: %s", self.rank, exc)
+                try:
+                    sentinel.reset_device()
```
[the remark this answers is not in the document]
Comment (205150940) [reply vllm_ascend/worker/sentinel/npu_worker_sentinel.py:64]: 并未出现hang住情况，无需增加传参

File: vllm_ascend/worker/sentinel/npu_worker_sentinel.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    def query_mask(self, ft_request: FaultToleranceRequest) -> dict:
+        """Report the dead-rank mask (upstream convention: 0=live, 1=dead)."""
+        return {"mask": get_ep_all2all_manager().query_active_mask().tolist()}
+
+    def reset_device(self) -> None:
+        NPUPlatform.set_device(self.device)
+        torch_npu.npu.stop_device(self.device.index)
+        torch_npu.npu.restart_device(self.device.index)
+        torch_npu.distributed.reinit_process_group(None, False)
+        torch.npu.synchronize()
```
[the remark this answers is not in the document]
Comment (205150940) [reply vllm_ascend/worker/sentinel/npu_worker_sentinel.py:97]: 同上，无需增加配置项

File: vllm_ascend/worker/model_runner_v1.py [hunk 3]
```diff
@@ -2080,6 +2080,25 @@ def take_draft_token_ids(self) -> DraftTokenIds | None:
...
+            yield
+            return
+
+        self.prepare_inputs_event.synchronize()
+        try:
+            yield
+        except Exception as e:
+            raise RuntimeError(
+                "prepare_inputs_event record skipped due to fault."
+            ) from e
```
[the remark this answers is not in the document]
Comment (205150940) [reply vllm_ascend/worker/model_runner_v1.py:2152]: 此处需要手动抛出异常，避免进程hang住

File: vllm_ascend/distributed/device_communicators/npu_communicator.py [hunk 4]
```diff
@@ -21,20 +21,83 @@
...
+        self._device = device
+        self._dead: set[int] = set()
+        self._num_local_experts: int = 0
+
+        # elastic_info layout: [is_scaling_down, dense ep world size,
+        #  shared_expert_rank_num, num_physical_experts] + table1(orig->dense)
+        #  + table2(dense->orig). num_physical_experts is derived from the
+        #  dead set and num_local_experts on every rebuild.
+        size = 4 + 2 * ep_world_size
+        self._elastic_info_host = torch.zeros(size, dtype=torch.int32)
```
[the remark this answers is not in the document]
Comment (205150940) [reply vllm_ascend/distributed/device_communicators/npu_communicator.py]: No changes needed.
